### DISCUSSION TITLE
移植Linux自带的包管理方式

### DISCUSSION THREAD

**Discussion by playgithub** [xmake-io/xmake#1607] [Ideas]
一点想法，听听意见

Linux自带的包管理特点
* 安装默认依赖很方便，支持传递依赖递归式安装
* 如果需要定制，可以安装到`/usr`外目录，如`/usr/local`或`/opt`，cmake `find_package`优先找这个目录就行

有没有办法做成跨平台？
我想了一下，主要两个难点
1. 传递依赖，比如，A依赖B，B依赖C，安装A的时候自动安装C
2. 二进制可能不存在

解决办法
第一个问题，提供依赖文件
第二个问题，源码级支持

有点类似vcpkg，但是库的查找路径统一。如果用cmake，`CMakeLists.txt`不需要额外修改；如果用.sln，设置几个目录就可以。

**Comment by waruqi** [comment]
没懂，这是要表达什么呢？ 

> 安装默认依赖很方便，支持传递依赖递归式安装
如果需要定制，可以安装到/usr外目录，如/usr/local或/opt，cmake find_package优先找这个目录就行
有没有办法做成跨平台？

xmake 的包管理，这些原本就全支持了

> 解决办法
第一个问题，提供依赖文件
第二个问题，源码级支持

这些原本也全支持

**Reply by playgithub** [reply]
都支持的就太好了，我先看看文档

**Reply by playgithub** [reply]
看文档 https://xmake.io/#/package/remote_package?id=cmake
xmake可以调用cmake

有些问题，举例
要安装`poppler`
`poppler`依赖`cairo`，`cairo`依赖`zlib`
* 问题1
  是不是要求`poppler` `cairo` `zlib`都有xmake project？
* 问题2
  如果安装成功，xmake是否透明，也就是等价于一个一个手动安装到比如`/usr/local`或`/opt`，没有附带任何xmake相关的依赖？

**Reply by waruqi** [reply]
> 是不是要求poppler cairo zlib都有xmake project？

不需要，有啥就用啥，可以 port 到 xmake.lua 对接，也可以直接对接 cmake/autotools/meson/scons的库。

https://github.com/xmake-io/xmake-repo/blob/master/packages/p/poppler/xmake.lua

> 如果安装成功，xmake是否透明，也就是等价于一个一个手动安装到比如/usr/local或/opt，没有附带任何xmake相关的依赖？

装在独立目录，会自动附带所有依赖的 links , packages, defines，用户不用关心。。

我建议你先用一遍，自己体验下 不就全知道了么

**Comment by playgithub** [comment]
安装了xmake，想安装`poppler`，看看安装后布局，如下报错，应该怎么安装？
```
$ xmake install poppler
note: xmake.lua not found, try generating it ...
```

期望的结果是，上面的命令以默认配置装好`poppler`以及所有依赖（包括传递依赖）
如果有些依赖xmake不能识别，可以自己增加到xmake
如果有些依赖需要特别配置，可以override配置，或者修改现有配置

**Reply by waruqi** [reply]
仔细看文档哈，都有说明，readme 里面都有基础用法

编译集成走 xmake，单独安装走 xrepo https://xrepo.xmake.io/#/zh-cn/getting_started

**Reply by waruqi** [reply]
> 如果有些依赖需要特别配置，可以override配置，或者修改现有配置

不都支持么，仔细看文档

**Reply by playgithub** [reply]
什么情况？
```
$ xrepo install poppler
updating repositories .. ok
error: package(freetype/>=2.9:{}): conflict dependences with package(freetype/latest:{"bzip2=false","png=false","woff2=false"})!
error: execv(xmake require -j 24 --extra={system=false} poppler) failed(255)
```

**Reply by waruqi** [reply]
> error: package(freetype/>=2.9:{}): conflict dependences with package(freetype/latest:{"bzip2=false","png=false","woff2=false"})!

包有点问题，具体等这个 patch merge后，再试下 https://github.com/xmake-io/xmake-repo/pull/597

**Reply by playgithub** [reply]
这个问题不大，对xmake质量有信心。
先了解以下设计，感觉有点像conan，如果一个普通工程要用`xmake`安装的库，还是能感觉到xmake的存在，如果感觉不到就好了，就像Linux下使用自带的包管理工具安装的库一样。
也就是说，安装结果等同于手动安装每个库以及依赖，比手动的优点很明显：
* 自动解决所有依赖
* 支持源码级安装
* 跨平台
* 干净卸载

这样的话，工程用最基本的方式就可以支持使用这些库，不需要引入额外的概念，比如
* `cmake -DCMAKE_PREFIX_PATH=/path/to/my_libs_root`
* Visual Studio，添加`include path` `link directory path` `link libs`等就可以

**Reply by waruqi** [reply]
目前 xmake/xrepo 安装的包 主要用于 xmake 项目自身的集成和使用，原本就是给 xmake 用的，xrepo 也是方便用户使用 xmake 提供的独立包管理器。

暂时不考虑对外输出给 cmake/meson 尽管理论上都可以支持，但是人力不够，我只能优先支持 xmake 自身项目集成。

即使是 conan，也需要 cmakelists.txt 里面注入一段配置才能集成支持它的包。

**Reply by waruqi** [reply]
如果想要 cmake 和 vs无缝支持 xmake 装的包，只有给他们提 patch ，让他们自己原生支持。。你可以尝试给 cmake 提 pr ，直接改进 find_package 从 xmake 里面找到包。

**Reply by waruqi** [reply]
> 什么情况？
> 
> ```
> $ xrepo install poppler
> updating repositories .. ok
> error: package(freetype/>=2.9:{}): conflict dependences with package(freetype/latest:{"bzip2=false","png=false","woff2=false"})!
> error: execv(xmake require -j 24 --extra={system=false} poppler) failed(255)
> ```

更新下仓库再试试

```console
xrepo update-repo
xrepo install poppler
```

**Reply by playgithub** [reply]
试了，可以了

**Comment by playgithub** [comment]
`xmake install zlib`成功
安装在`~/.xmake/packages/z/zlib/1.2.11/832cf454431442ab9f0ec95e35aac34d`
这样普通`CMakeLists.txt`找不到
如果手动安装`zlib`，默认会安装到`/usr/local`下，默认能被`find_package`找到，`xmake`能做到这个效果就好了

**Reply by waruqi** [reply]
> `xmake install zlib`成功
> 安装在`~/.xmake/packages/z/zlib/1.2.11/832cf454431442ab9f0ec95e35aac34d`
> 这样普通`CMakeLists.txt`找不到
> 如果手动安装`zlib`，默认会安装到`/usr/local`下，默认能被`find_package`找到，`xmake`能做到这个效果就好了

不会考虑，你自己也说了，要支持 特定配置重写，全局 /usr/local 只能放置一个 配置版本的 库。。。如果一个项目，多个target 同时依赖多个 不同配置的 zlib 库，怎么搞？互相覆盖么？

这不是 homebrew apt 。。只需安装一个库就行了，xmake 要可能要同时支持安装 zlib 的不同配置，比如 shared/static版本，debug版本，特定模块被启用的版本，还有不同版本号，有可能是 n 种组合，都放到 /usr/local/lib/libz.a 怎么玩？

xmake装的包，走 add_requires/add_packages集成就行了，如果要在cmake里面用，自己用 `xrepo env shell` 切到特定绑定对应包的环境 执行 cmake 就行了，具体看文档和readme说明 https://github.com/xmake-io/xrepo#enter-the-package-shell-environment

**Reply by playgithub** [reply]
> 不会考虑，你自己也说了，要支持 特定配置重写，全局 /usr/local 只能放置一个 配置版本的 库。。。如果一个项目，多个target 同时依赖多个 不同配置的 zlib 库，怎么搞？互相覆盖么？
> 这不是 homebrew apt 。。只需安装一个库就行了，xmake 要可能要同时支持安装 zlib 的不同配置，比如 shared/static版本，debug版本，特定模块被启用的版本，还有不同版本号，有可能是 n 种组合，都放到 /usr/local/lib/libz.a 怎么玩？
> xmake装的包，走 add_requires/add_packages集成就行了，如果要在cmake里面用，自己用 `xrepo env shell` 切到特定绑定对应包的环境 执行 cmake 就行了，具体看文档和readme说明 https://github.com/xmake-io/xrepo#enter-the-package-shell-environment

要配置组合灵活，还是用哈希目录最好。
**`xrepo env shell`可以做到让原生cmake工程不需要任何修改就可以发现并使用`xrepo`安装的库吗？**
那比`vcpkg`和`conan`都先进，准备环境而不是修改原生工程，我觉得这才是正确的做法。

**Reply by waruqi** [reply]
> xrepo env shell可以做到让原生cmake工程不需要任何修改就可以发现并使用xrepo安装的库吗？

理论上完全可以，但目前这个特性 用的人不多，也许测试不全，你可以试试，有坑可以提 issues 改进

也许一些包安装的时候，结构和 pc 文件变动，或者缺失 xx.cmake，导致 cmake find_package找不到也有可能的，遇到这种，需要改进下包，可以提 pr 到 xmake-repo 改进

毕竟这依赖 cmakelists.txt 自身项目，很多 cmake 项目维护的不成熟，找不到也正常，所以尽量用 xmake 项目维护

**Reply by playgithub** [reply]
好，环境设置好还找不到，那是那些包的问题或者小BUG，不是问题，这个设计绝对比老外的先进

**Comment by xq114** [comment]
cmake查找包的方式有两种：
一个FindXXX.cmake，这一种需要你自己写一个文件放在cmake/module里面才能找到，直接find_package(xxx)是找不到的，不使用cmake作为构建工具的大部分package属于这种；另一个XXXConfig.cmake，这种需要包本身有导出config文件，只要激活了xrepo env shell虚拟环境就可以找到，opencv等属于这种。

cmake还内置了少数package脚本，既不需要FindXXX.cmake也不需要XXXConfig.cmake，激活xrepo env shell之后直接find_package就能找到，zlib、libpng等属于这种。

**Comment by playgithub** [comment]
`conan`的`cmake-find-package-generator`可以为每个依赖的`conan package`生成`Find<package_name>.cmake`
https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_find_package_generator.html#cmake-find-package-generator
不需要修改`CMakeLists.txt`

**Reply by xq114** [reply]
https://github.com/xmake-io/xmake/issues/1418

**Reply by waruqi** [reply]
还没时间搞，谁有兴趣，可以帮忙搞下。

**Comment by playgithub** [comment]
#1418 说下一个版本弄，`conan 2.0`有新做法了，参考一下`CMakeToolchain`https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#cmaketoolchain

**Reply by waruqi** [reply]
> #1418 说下一个版本弄，`conan 2.0`有新做法了，参考一下`CMakeToolchain`https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#cmaketoolchain

当初说的下个版本，其实已经拖了一个版本，近期估计没时间搞了，先放着吧

**Reply by xq114** [reply]
这个conantoolchain和包管理有什么关系？这只是一个cmake的preset

**Reply by playgithub** [reply]
> 这个conantoolchain和包管理有什么关系？这只是一个cmake的preset

`conan`的`generator`设置为`CMakeToolchain`，然后会根据`conanfile.txt`里面的依赖生成`conantoolchain.cmake`，
调用`cmake`是加上`-DCMAKE_TOOLCHAIN_FILE=conantoolchain.cmake`，`cmake`就能找到`conanfile.txt`里面的库了，不需要修改原生工程。`conan`文档一般，还没试过。

**Reply by playgithub** [reply]
> 这个conantoolchain和包管理有什么关系？这只是一个cmake的preset

`CMakeToolchain`验证了，可以在不修改原生工程的情况下build，需要配合`CMakeDeps`

官方例子(https://github.com/conan-io/examples/tree/master/libraries/poco/md5) 改了一下

`conanfile.txt`
```ini
[requires]
poco/1.11.0

[generators]
CMakeDeps
CMakeToolchain
```

`CMakeLists.txt`
```cmake
cmake_minimum_required(VERSION 2.8.12)

project(MD5Encrypter)

find_package(Poco REQUIRED)

if(CMAKE_VERSION VERSION_LESS 3.0.0)
    include(CheckCXXCompilerFlag)
    check_cxx_compiler_flag(-std=c++11 COMPILER_SUPPORTS_CXX11)
    check_cxx_compiler_flag(-std=c++0x COMPILER_SUPPORTS_CXX0X)
    if(COMPILER_SUPPORTS_CXX11)
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
    elseif(COMPILER_SUPPORTS_CXX0X)
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x")
    endif()
else()
    SET(CMAKE_CXX_STANDARD 11)
    SET(CMAKE_CXX_STANDARD_REQUIRED ON)
endif()

add_executable(md5 md5.cpp)

target_link_libraries(md5 PRIVATE Poco::Poco)
```

`md5.cpp`
```cpp
#include "Poco/MD5Engine.h"
#include "Poco/DigestStream.h"

#include <iostream>


int main(int argc, char** argv)
{
    Poco::MD5Engine md5;
    Poco::DigestOutputStream ds(md5);
    ds << "abcdefghijklmnopqrstuvwxyz";
    ds.close();
    std::cout << Poco::DigestEngine::digestToHex(md5.digest()) << std::endl;
    return 0;
}
```

```shell
$ mkdir out
$ cd out
$ conan install ..
$ cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake -DCMAKE_BUILD_TYPE=release
$ cmake --build .
$ ./md5 
c3fcd3d76192e4007dfb496cca67e13b
```
